rllib/core/distribution/torch/torch_distribution.py

```diff
@@ -0,0 +1,707 @@
+"""The main difference between this and the old ActionDistribution is that this one
+has more explicit input args. So that the input format does not have to be guessed from
+the code. This matches the design pattern of torch distribution which developers may
+already be familiar with.
+"""
+import gymnasium as gym
+import numpy as np
+from typing import Dict, Iterable, List, Optional
+import tree
+import abc
+
+
+from ray.rllib.models.distributions import Distribution
+from ray.rllib.utils.annotations import override, DeveloperAPI
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.numpy import MAX_LOG_NN_OUTPUT, MIN_LOG_NN_OUTPUT, SMALL_NUMBER
+from ray.rllib.utils.typing import TensorType, Union, Tuple
+
+torch, nn = try_import_torch()
+
+
+@DeveloperAPI
+class TorchDistribution(Distribution, abc.ABC):
+    """Wrapper class for torch.distributions."""
+
+    def __init__(self, *args, **kwargs):
+        super().__init__()
+        self._dist = self._get_torch_distribution(*args, **kwargs)
+
+    @abc.abstractmethod
+    def _get_torch_distribution(
+        self, *args, **kwargs
+    ) -> "torch.distributions.Distribution":
+        """Returns the torch.distributions.Distribution object to use."""
+
+    @override(Distribution)
+    def logp(self, value: TensorType, **kwargs) -> TensorType:
+        return self._dist.log_prob(value, **kwargs)
+
+    @override(Distribution)
+    def entropy(self) -> TensorType:
+        return self._dist.entropy()
+
+    @override(Distribution)
+    def kl(self, other: "Distribution") -> TensorType:
+        return torch.distributions.kl.kl_divergence(self._dist, other._dist)
+
+    @override(Distribution)
+    def sample(
+        self,
+        *,
+        sample_shape=None,
+    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
+        sample = self._dist.sample(
+            sample_shape if sample_shape is not None else torch.Size()
+        )
+        return sample
+
+    @override(Distribution)
+    def rsample(
+        self,
+        *,
+        sample_shape=None,
+    ) -> Union[TensorType, Tuple[TensorType, TensorType]]:
+        rsample = self._dist.rsample(
+            sample_shape if sample_shape is not None else torch.Size()
+        )
+        return rsample
+
+    @classmethod
+    @override(Distribution)
+    def from_logits(cls, logits: TensorType, **kwargs) -> "TorchDistribution":
+        return cls(logits=logits, **kwargs)
+
+
+@DeveloperAPI
+class TorchCategorical(TorchDistribution):
```

Comment:
Why do we override the whole distribution file?

Reply:
Oh, we moved it from the old-stack `models` directory to the new-stack `core/distribution` directory.